python/demo/gsoc2022/demo_sbc.py

```diff
@@ -0,0 +1,244 @@
+radius_dom = 1 * um
+
+# the smaller the mesh_factor, the finer is the mesh
+mesh_factor = 1.2
+
+# finite element degree
+degree = 3
+
+# wavelength sweep
+wl0 = 0.4 * um
+
+# background refractive index
+n_bkg = 1.33
+eps_bkg = n_bkg**2
+
+# mesh size inside the wire
+in_wire_size = mesh_factor * 7 * nm
+
+# mesh size at the boundary of the wire
+on_wire_size = mesh_factor * 3 * nm
+
+# mesh size in the vacuum
+bkg_size = mesh_factor * 60 * nm
+
+# mesh size at the boundary
+boundary_size = mesh_factor * 30 * nm
+
+# tags for the subdomains
+au_tag = 1          # gold wire
+bkg_tag = 2         # background
+boundary_tag = 3    # boundary
+
+# mesh definition in gmsh
+
+gmsh.initialize(sys.argv)
+if MPI.COMM_WORLD.rank == 0:
+
+    gmsh.model.add("nanowire")
+
+    # a dummy boundary is added for setting a finer mesh
+    gmsh.model.occ.addCircle(0, 0, 0, radius_wire * 0.8, angle1=0, angle2=2 * pi, tag=1)
+    gmsh.model.occ.addCircle(0, 0, 0, radius_wire, angle1=0, angle2=2 * pi, tag=2)
+
+    # a dummy boundary is added for setting a finer mesh
+    gmsh.model.occ.addCircle(0, 0, 0, radius_dom * 0.9, angle1=0, angle2=2 * pi, tag=3)
+    gmsh.model.occ.addCircle(0, 0, 0, radius_dom, angle1=0, angle2=2 * pi, tag=4)
+
+    gmsh.model.occ.addCurveLoop([1], tag=1)
+    gmsh.model.occ.addPlaneSurface([1], tag=1)
+
+    gmsh.model.occ.addCurveLoop([2], tag=2)
+    gmsh.model.occ.addCurveLoop([1], tag=3)
+    gmsh.model.occ.addPlaneSurface([2, 3], tag=2)
+
+    gmsh.model.occ.addCurveLoop([3], tag=4)
+    gmsh.model.occ.addCurveLoop([2], tag=5)
+    gmsh.model.occ.addPlaneSurface([4, 5], tag=3)
+
+    gmsh.model.occ.addCurveLoop([4], tag=6)
+    gmsh.model.occ.addCurveLoop([3], tag=7)
+    gmsh.model.occ.addPlaneSurface([6, 7], tag=4)
+
+    gmsh.model.occ.synchronize()
+
+    gmsh.model.addPhysicalGroup(2, [1, 2], tag=au_tag)
+    gmsh.model.addPhysicalGroup(2, [3, 4], tag=bkg_tag)
+
+    gmsh.model.addPhysicalGroup(1, [4], tag=boundary_tag)
+
+    gmsh.model.mesh.setSize([(0, 1)], size=in_wire_size)
+    gmsh.model.mesh.setSize([(0, 2)], size=on_wire_size)
+    gmsh.model.mesh.setSize([(0, 3)], size=bkg_size)
+    gmsh.model.mesh.setSize([(0, 4)], size=boundary_size)
+
+    gmsh.model.mesh.generate(2)
+
+mesh, cell_tags, facet_tags = gmsh_model_to_mesh(gmsh.model, cell_data=True, facet_data=True, gdim=2)
+
+MPI.COMM_WORLD.barrier()
+
+# definition of finite element for the electric field
+curl_el = ufl.FiniteElement("N1curl", mesh.ufl_cell(), 3)
+
+# definition of finite element for the r function (see next)
+lagr_el = ufl.FiniteElement("CG", mesh.ufl_cell(), 2)
+
+# function space for the electric field
+V = fem.FunctionSpace(mesh, curl_el)
+
+# wavevector of the background field
+k0 = 2 * np.pi / wl0
+
+# angle of incidence of the background field with respect to the horizontal axis
+deg = np.pi / 180
+theta = 45 * deg
+
+# plane wave function
+f = background_electric_field(theta, n_bkg, k0)
+Eb = fem.Function(V, dtype=np.complex128)
```

Comment:
Instead of hardcoding `dtype=np.complex128`  I think we should add 
```python
if not np.issubdtype(PETSc.ScalarType, np.complexfloating):
    print("Demo should only be executed with DOLFINx complex mode")
    exit(0)
```
at the start of the demo.